File: pufferlib/ocean/artillery/artillery.c
#include <time.h>
#include "artillery.h"
#include "puffernet.h"

void allocate(Artillery* env) {
    init(env);
    int obs_size = 11;
    env->observations = (float*)calloc(obs_size, sizeof(float));
    env->actions = (int*)calloc(2, sizeof(int));
    env->rewards = (float*)calloc(1, sizeof(float));
    env->terminals = (unsigned char*)calloc(1, sizeof(unsigned char));
}

void free_allocated(Artillery* env) {
    free(env->actions);
    free(env->observations);
    free(env->terminals);
    free(env->rewards);
    c_close(env);
}

void demo() {
    Artillery env = {
        .adj = 0.014364991132735087,
        .dist_fade = 0.11107816724558334,
        .frameskip = 1,
        .miss_penalty = -0.05521742600140713,
        .max_reward_dist = 44.700659812685586,
        .max_dist0 = 104.59870905085756,
        .out_bounds_penalty = -0.1,
        .turn_penalty = -0.1,
        .turn_penalty_delay = 98,
        .render = 1,
    };
    allocate(&env);

    env.client = make_client(&env);

    Weights* weights = load_weights("resources/artillery/puffer_artillery_weights.bin", 135051);
    int logit_sizes[2] = {5, 5};
    int obs_size = 11;
    LinearLSTM* net = make_linearlstm(weights, 1, obs_size, logit_sizes, 2);

    c_reset(&env);
    SetTargetFPS(30);
    while (!WindowShouldClose()) {
        if (IsKeyDown(KEY_LEFT_SHIFT)) {
            env.actions[0] = 4.0f;
            env.actions[1] = 4.0f;

            if (IsKeyDown(KEY_SPACE)) env.actions[0] = 0;
            if (IsKeyDown(KEY_W)) env.actions[0] = 1;
            if (IsKeyDown(KEY_S)) env.actions[0] = 2;
            if (IsKeyDown(KEY_A)) env.actions[0] = 3;
            if (IsKeyDown(KEY_D)) env.actions[0] = 4;

            if (IsKeyDown(KEY_ENTER)) env.actions[1] = 0;
            if (IsKeyDown(KEY_UP)) env.actions[1] = 1;
            if (IsKeyDown(KEY_DOWN)) env.actions[1] = 2;
            if (IsKeyDown(KEY_LEFT)) env.actions[1] = 3;
            if (IsKeyDown(KEY_RIGHT)) env.actions[1] = 4;
        } else {
            int discrete_actions[2];
            forward_linearlstm(net, env.observations, discrete_actions);
            env.actions[0] = discrete_actions[0];
            env.actions[1] = discrete_actions[1];
        }
        c_step(&env);
        c_render(&env);
    }
    free_linearlstm(net);
    free(weights);
    free_allocated(&env);
    close_client(env.client);
}

int main() {
    demo();
}

File: pufferlib/ocean/artillery/artillery.py
import numpy as np
import gymnasium
import time

import pufferlib
from pufferlib.ocean.artillery import binding

class Artillery(pufferlib.PufferEnv):
    def __init__(self, num_envs=1, render_mode=None,
                 frameskip=1,
                 min_aim_angle=0.56, max_aim_angle=1.56, max_reward_dist=45,
                 adj=0.0144, dist_fade=0.11, turn_penalty_delay=98, max_dist0=105,
                 turn_penalty=-0.1, miss_penalty=-0.055, render=1, out_bounds_penalty=-0.1,
                 log_interval=128,
                 seed=7,
                 buf=None):
        obs_size = 11
        self.single_observation_space = gymnasium.spaces.Box(low=0, high=1, shape=(obs_size,), dtype=np.float32)
        self.render_mode = render_mode
        self.num_agents = num_envs
        self.log_interval = log_interval
        self.tick = 0

        self.single_action_space = gymnasium.spaces.MultiDiscrete([5, 5])

        super().__init__(buf)

        self.actions = self.actions

        self.c_envs = binding.vec_init(
            self.observations, self.actions, self.rewards, self.terminals, self.truncations, num_envs,
            seed, num_envs=num_envs, seed=seed, frameskip=frameskip,
            min_aim_angle=min_aim_angle, max_aim_angle=max_aim_angle, max_reward_dist=max_reward_dist,
            adj=adj, dist_fade=dist_fade, turn_penalty_delay=turn_penalty_delay, max_dist0=max_dist0,
            turn_penalty=turn_penalty, miss_penalty=miss_penalty, render=render,
            out_bounds_penalty=out_bounds_penalty
        )

    def reset(self, seed=0):
        binding.vec_reset(self.c_envs, seed)
        self.tick = 0
        return self.observations, []

    def step(self, actions):
        self.actions[:] = actions

        self.tick += 1
        binding.vec_step(self.c_envs)

        info = []
        if self.tick % self.log_interval == 0:
            info.append(binding.vec_log(self.c_envs))

        return (self.observations, self.rewards,
            self.terminals, self.truncations, info)

    def render(self):
        binding.vec_render(self.c_envs, 0)

    def close(self):
        binding.vec_close(self.c_envs)

def test_performance(timeout=10, atn_cache=1024):
    env = Artillery(num_envs=1)
    env.reset()
    tick = 0

    actions = np.random.randint(0, 5, (atn_cache, env.num_agents, 2))

    import time
    start = time.time()
    while time.time() - start < timeout:
        atn = actions[tick % atn_cache]
        env.step(atn)
        tick += 1

if __name__ == '__main__':
    test_performance()
